### PR TITLE
plan: enhance predicate pushdown over join

### PR DESCRIPTION
**Pull request by eurekaka**
Derive superset of DNF predicates and push it to children
plan nodes as much as possible.

### What problem does this PR solve? 

fix https://github.com/pingcap/tidb/issues/7628

This PR should have performance improvement for TPC-H Q7 and Q19 at least.

before this PR:
```
drop table if exists t1;
drop table if exists t2;

create table t1(id int primary key, a int, b int, c int);
create table t2(id int primary key, a int, b int, c int);

mysql> explain select * from t1 join t2 on t1.b = t2.b where (t1.a=1 and t2.a=1) or (t1.a=2 and t2.a=2);
+-----------------------+----------+------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| id                    | count    | task | operator info                                                                                                                                                       |
+-----------------------+----------+------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| HashLeftJoin_7        | 12500.00 | root | inner join, inner:TableReader_12, equal:[eq(test.t1.b, test.t2.b)], other cond:or(and(eq(test.t1.a, 1), eq(test.t2.a, 1)), and(eq(test.t1.a, 2), eq(test.t2.a, 2))) |
| ├─TableReader_10      | 10000.00 | root | data:TableScan_9                                                                                                                                                    |
| │ └─TableScan_9       | 10000.00 | cop  | table:t1, range:[-inf,+inf], keep order:false, stats:pseudo                                                                                                         |
| └─TableReader_12      | 10000.00 | root | data:TableScan_11                                                                                                                                                   |
|   └─TableScan_11      | 10000.00 | cop  | table:t2, range:[-inf,+inf], keep order:false, stats:pseudo                                                                                                         |
+-----------------------+----------+------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

no predicate is pushed down because the filter is in DNF and involves columns from different tables;

after this PR:
```
mysql> explain select * from t1 join t2 on t1.b = t2.b where (t1.a=1 and t2.a=1) or (t1.a=2 and t2.a=2);
+-------------------------+----------+------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| id                      | count    | task | operator info                                                                                                                                                       |
+-------------------------+----------+------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| HashLeftJoin_7          | 25.00    | root | inner join, inner:TableReader_14, equal:[eq(test.t1.b, test.t2.b)], other cond:or(and(eq(test.t1.a, 1), eq(test.t2.a, 1)), and(eq(test.t1.a, 2), eq(test.t2.a, 2))) |
| ├─TableReader_11        | 20.00    | root | data:Selection_10                                                                                                                                                   |
| │ └─Selection_10        | 20.00    | cop  | or(eq(test.t1.a, 1), eq(test.t1.a, 2))                                                                                                                              |
| │   └─TableScan_9       | 10000.00 | cop  | table:t1, range:[-inf,+inf], keep order:false, stats:pseudo                                                                                                         |
| └─TableReader_14        | 20.00    | root | data:Selection_13                                                                                                                                                   |
|   └─Selection_13        | 20.00    | cop  | or(eq(test.t2.a, 1), eq(test.t2.a, 2))                                                                                                                              |
|     └─TableScan_12      | 10000.00 | cop  | table:t2, range:[-inf,+inf], keep order:false, stats:pseudo                                                                                                         |
+-------------------------+----------+------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

new filters in `Selection_13` and `Selection_10` is derived from the original filters and pushed down.

### What is changed and how it works?

rewrite original filter `(t1.a=1 and t2.a=1) or (t1.a=2 and t2.a=2)` into `((t1.a=1 and t2.a=1) or (t1.a=2 and t2.a=2)) AND (t1.a=1 or t1.a=2) AND (t2.a=1 or t2.a=2)`, and the newly derived CNF items can be pushed down.

Outer join is supported as well.

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported function/method change

Side effects

 - Increased code complexity

Related changes

 - Need to be included in the release note: `enhance predicate pushdown over join operator`

### CODE REVIEW

**Comment by eurekaka** [comment]
@zz-jason @winoros PTAL

File: plan/logical_plan_builder.go [hunk 1]
```diff
@@ -208,12 +211,50 @@ func extractOnCondition(conditions []expression.Expression, left LogicalPlan, ri
...
+			}
 			otherCond = append(otherCond, expr)
 		}
 	}
 	return
 }
 
+// deriveOtherConditions given a LogicalJoin, check the OtherConditions to see if we can derive more
+// conditions for left/right child pushdown.
+func deriveOtherConditions(p *LogicalJoin, deriveLeft bool, deriveRight bool) (leftCond []expression.Expression,
```
Comment (zz-jason) [inline plan/logical_plan_builder.go]: move this function to "plan/rule_predicate_push_down.go"?

File: expression/util.go [hunk 2]
```diff
@@ -372,6 +372,46 @@ func extractFiltersFromDNF(ctx sessionctx.Context, dnfFunc *ScalarFunction) ([]E
...
+					newCNFItems = append(newCNFItems, relaxedCNFItem)
+				}
+				// If relaxed expression for embedded DNF is univeral set, just drop this CNF item
+				continue
+			}
+			// This cnfItem must be simple expression now
+			if ExprFromSchema(cnfItem, schema) {
+				newCNFItems = append(newCNFItems, cnfItem)
+			}
+			// If simple cnfItem cannot be fully covered by schema, just drop this CNF item
```
Comment (zz-jason) [inline expression/util.go]: move this comment together with line 400?

**Comment by winoros** [comment]
Have we tested this branch's performance on TPCH Q7 and Q8?

**Comment by eurekaka** [comment]
@winoros not yet, will provide result later.

**Comment by winoros** [comment]
I come up with the question what will there be performance regression when the pushed one's selectivity is quite big?
Since this transformation won't remove the original filter so the filter will be calculated multiple times.

File: plan/logical_plan_test.go [hunk 3]
```diff
@@ -439,6 +440,159 @@ func (s *testPlanSuite) TestPredicatePushDown(c *C) {
...
+		rightPlan, ok := join.children[1].(*DataSource)
+		c.Assert(ok, IsTrue, comment)
+		leftCond := fmt.Sprintf("%s", leftPlan.pushedDownConds)
+		rightCond := fmt.Sprintf("%s", rightPlan.pushedDownConds)
+		c.Assert(leftCond, Equals, ca.left, comment)
+		c.Assert(rightCond, Equals, ca.right, comment)
+	}
+}
+
+func (s *testPlanSuite) TestOuterWherePredicatePushDown(c *C) {
```
[the remark this answers is not in the document]
Comment (eurekaka) [reply plan/logical_plan_test.go]: Hmm, we need to test the predicates are correctly derived and pushed to specific child in `PredicatePushDown` as well, so I prefer building logical plan.

**Comment by zz-jason** [comment]
An ideal solution is to calculate the selectivity of the to-be-pushed predicates, decide whether to push that predicate according to the calculated selectivity. But this selectivity might not be so accurate. 

It seems that the adaptive execution plan adopted in oracle optimizer is more suitable. They just constructed two plans, one has the pushed predicates while the other doesn't. And using the statistics collector to record the number of rows produced by the child operator during the execution phase. Deciding whether to use the pushed predicates according to the number of rows observed by the statistics collector.

**Pull request merged**